src/node/interfaces.cpp

```diff
@@ -182,38 +182,51 @@ class NodeImpl : public Node
     }
     int getNumBlocks() override
     {
-        LOCK(::cs_main);
-        assert(std::addressof(::ChainActive()) == std::addressof(m_context->chainman->ActiveChain()));
-        return m_context->chainman->ActiveChain().Height();
+        if (m_context->chainman) {
+            LOCK(::cs_main);
+            assert(std::addressof(::ChainActive()) == std::addressof(m_context->chainman->ActiveChain()));
+            return m_context->chainman->ActiveChain().Height();
+        }
+        return 0;
     }
     uint256 getBestBlockHash() override
     {
-        assert(std::addressof(::ChainActive()) == std::addressof(m_context->chainman->ActiveChain()));
-        const CBlockIndex* tip = WITH_LOCK(::cs_main, return m_context->chainman->ActiveChain().Tip());
+        const CBlockIndex* tip;
+        if (m_context->chainman) {
```

Comment:
In commit "node/interfaces: Meaningful locking and avoid potential UB" (5de3e3609ab591d88281558182efdec39b065541)

> Same here. You don't need this local variable if you exit early when `chainman` doesn't exist.

I think John's suggestions are good, but Marco's suggestion to use an accessor that asserts https://github.com/bitcoin/bitcoin/pull/21391#discussion_r596601383 would be even better. That way we don't need to waste brain cells thinking about what values to return if chainman is null. Like if getBestBlockHash returns the genesis block hash in that case should getNumBlocks return 1 instead of 0 to be consistent? Easier to take everything down and assert.

Reply:
Seems reasonable to me!

In general, I like the concept of making all components optional, since it enforces decoupling. However, it's hard to imagine what a bitcoind without chainman would look like.